### PR TITLE
fix: Fastify.listen({path: '...'}) should not also listen on a port

### LINKED ISSUE #4007: Server will also listen on a random port when calling app.listen({path: 'foo.sock'})
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.0.2

### Plugin version

_No response_

### Node.js version

16.15.1

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.4

### Description

If I pass to `listen` only a `path` to listen on, the server will also listen on a random port.

I think this happens due to this code here: https://github.com/fastify/fastify/blob/78adc9cc76f26ad2461d15103452f64487d806df/lib/server.js#L297-L299

P.S. I got a deprecation warning even for `app.listen(option, (err) => { ... })` which AFAIK should not be deprecated.

### Steps to Reproduce

```js
const Fastify = require('fastify');
const app = Fastify({logger: true});
app.listen({ path: 'foo.sock' });
```

### Expected Behavior

It only listens on `path`.

### PR DESCRIPTION
**Pull request by segevfiner**
Fixes #4007

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/server.js [hunk 1]
```diff
@@ -41,7 +41,7 @@ function createServer (options, httpHandler) {
       listenOptions.cb = cb
     }
 
-    const { host = 'localhost' } = listenOptions
+    const { host = listenOptions.path == null ? 'localhost' : null } = listenOptions
```
Comment (jsumners) [inline lib/server.js:44]:
```suggestion
    // <insert link to GH issue here>
    const { host = listenOptions.path == null ? 'localhost' : null } = listenOptions
```

This is a little difficult to reason through. The one line change is clever, but I think we should expand it out into a full `if` conditional with a comment or two to explain why this is being done.

**Review by jsumners — CHANGES_REQUESTED** [review]

File: lib/server.js (hunk 1 above)
Comment (segevfiner) [reply lib/server.js]: Let me know if you think there is some clearer way to write this.

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Review by Eomm — APPROVED** [review]

**Pull request merged**